shell2http.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "SHELL2HTTP" "" "December 2024" "" ""
HTTP\-server designed to execute shell commands\. It is suitable for development, prototyping or remote control, facilitates rapid iteration and testing of shell\-based functionalities\. Easy to set up using just two command\-line arguments: URL\-path and shell command\. Next, when an HTTP request is made for the specified path, the stdout of the shell script will be returned, that\'s all\.
.
.SH "Usage"
.
.nf

shell2http [options] /path "shell command" /path2 "shell command2" \.\.\.
options:
    \-host="host"      : host IP for http server (default bind to all interfaces)
    \-port=NNNN        : port for http server, 0 \- to receive a random port (default 8080)
    \-form             : parse query into environment vars, handle uploaded files
    \-form\-check       : regexp for check form fields (pass only vars that match the regexp)
    \-cgi              : run scripts in CGI\-mode:
                        \- set environment variables with HTTP\-request information
                        \- write POST|PUT|PATCH\-data to script STDIN (if is not set \-form)
                        \- parse headers from script (eg: "Location: URL\en\en")
    \-export\-vars=var  : export environment vars ("VAR1,VAR2,\.\.\.")
                        by default export PATH, HOME, LANG, USER, TMPDIR
    \-export\-all\-vars  : export all current environment vars
    \-no\-index         : don\'t generate index page
    \-add\-exit         : add /exit command
    \-log=filename     : log filename, default \- STDOUT
    \-shell="shell"    : shell for execute command, "" \- without shell (default "sh")
    \-cache=N          : caching command out for N seconds
    \-one\-thread       : run each shell command in one thread
    \-show\-errors      : show the standard output even if the command exits with a non\-zero exit code
    \-include\-stderr   : include stderr to output (default is stdout only)
    \-500              : return 500 error if shell exit code != 0
    \-cert=cert\.pem    : SSL certificate path (if specified \-cert/\-key options \- run https server)
    \-key=key\.pem      : SSL private key path
    \-basic\-auth=""    : setup HTTP Basic Authentication ("user_name:password"), can be used several times
    \-timeout=N        : set timeout for execute shell command (in seconds)
    \-no\-log\-timestamp : log output without timestamps
    \-version
    \-help
.
.fi
.
.P
In the \fB\-form\fR mode, variables are available for shell scripts:
.
.IP "\(bu" 4
$v_NNN \-\- data from query parameter with name "NNN" (example: \fBhttp://localhost:8080/path?NNN=123\fR)
.
.IP "\(bu" 4
$filepath_ID \-\- uploaded file path, ID \- id from \fB<input type=file name=ID>\fR, temporary uploaded file will be automatically deleted
.
.IP "\(bu" 4
$filename_ID \-\- uploaded file name from browser
.
.IP "" 0
.
.P
With \fB\-form\-check\fR option you can specify the regular expression for checking the form fields\. For example, if you want to allow only variables that contain the only digits, you can specify the following option: \fB\-form\-check=\'^[0\-9]+$\'\fR\. Then only requests like \fBhttp://localhost:8080/path?NNN=123\fR will be produce variable \fB$v_NNN\fR\.
.
.P
To setup multiple auth users, you can specify the \fB\-basic\-auth\fR option multiple times\. The credentials for basic authentication may also be provided via the \fBSH_BASIC_AUTH\fR environment variable\. You can specify the preferred HTTP\-method (via \fBMETHOD:\fR prefix for path): \fBshell2http GET:/date date\fR
.
.SH "Install"
MacOS:
.
.IP "" 4
.
.nf

brew install msoap/tools/shell2http
# update:
brew upgrade shell2http
.
.fi
.
.IP "" 0
.
.P
Download binaries from: releases \fIhttps://github\.com/msoap/shell2http/releases/latest\fR (MacOS/Linux/Windows/RaspberryPi)
.
.P
For Docker users, availeble tags see in Docker Hub \fIhttps://hub\.docker\.com/r/msoap/shell2http\fR:
.
.IP "" 4
.
.nf

docker pull msoap/shell2http
.
.fi
.
.IP "" 0
.
.P
Using snap (Ubuntu or any Linux distribution with snap):
.
.IP "" 4
.
.nf

# install stable version:
sudo snap install shell2http

# install the latest version:
sudo snap install \-\-edge shell2http

# update
sudo snap refresh shell2http
.
.fi
.
.IP "" 0
.
.P
\fINotice\fR: the snap\-package has its own sandbox with the \fB/bin\fR, \fB/usr/bin\fR directories which are not equal to system\-wide \fBPATH\fR directories and commands may not work as expected or not work at all\.
.
.P
Build from source (minimum Go version is 1\.12):
.
.IP "" 4
.
.nf

go install github\.com/msoap/shell2http@latest
# set link to your PATH if needed:
ln \-s $(go env GOPATH)/bin/shell2http ~/bin/shell2http
.
.fi
.
.IP "" 0
.
.P
Compile for MIPS CPU (for example, for some WiFi routers like Asus):
.
.IP "" 4
.
.nf

GOOS=linux GOARCH=mipsle GOMIPS=softfloat go build \-trimpath \-ldflags="\-w \-s" \-o shell2http \.
.
.fi
.
.IP "" 0
.
.SH "Examples"
.
.nf

shell2http /top "top \-l 1 | head \-10"
shell2http /date date /ps "ps aux"
shell2http \-export\-all\-vars /env \'printenv | sort\' /env/path \'echo $PATH\' /env/gopath \'echo $GOPATH\'
shell2http \-export\-vars=GOPATH /get \'echo $GOPATH\'
.
.fi
.
.SS "HTML calendar for current year"
.
.nf

shell2http /cal_html \'echo "<html><body><h1>Calendar</h1>Date: <b>$(date)</b><br><pre>$(cal $(date +%Y))</pre></body></html>"\'
.
.fi
.
.P
.
.SS "Using URL parameters (http://localhost:8080/form?from=10&to=100)"
.
.nf

shell2http \-form /form \'echo $v_from, $v_to\'
.
.fi
.
.P
.
.SS "CGI scripts"
.
.nf

shell2http \-cgi /user_agent \'echo $HTTP_USER_AGENT\'
shell2http \-cgi /set \'touch file; echo "Location: /another_path\en"\' # redirect
shell2http \-cgi /404 \'echo "Status: 404"; echo; echo "404 page"\' # custom HTTP code
.
.fi
.
.P
.
.SS "Upload file"
.
.nf

shell2http \-form \e
    GET:/form \'echo "<html><body><form method=POST action=/file enctype=multipart/form\-data><input type=file name=uplfile><input type=submit></form>"\' \e
    POST:/file \'cat $filepath_uplfile > uploaded_file\.dat; echo Ok\'
.
.fi
.
.P
Testing by upload file with curl:
.
.IP "" 4
.
.nf

curl \-i \-F uplfile=@some/file/path \'http://localhost:8080/file\'
.
.fi
.
.IP "" 0
.
.P
.
.SS "Simple http\-proxy server (for logging all URLs)"
Setup proxy as "http://localhost:8080/"
.
.IP "" 4
.
.nf

shell2http \-log=/dev/null \-cgi / \'echo $REQUEST_URI 1>&2; [ "$REQUEST_METHOD" == "POST" ] && post_param="\-d@\-"; curl \-sL $post_param "$REQUEST_URI" \-A "$HTTP_USER_AGENT"\'
.
.fi
.
.IP "" 0
.
.P
.
.SS "Test slow connection (http://localhost:8080/slow?duration=10)"
.
.nf

shell2http \-form /slow \'sleep ${v_duration:\-1}; echo "sleep ${v_duration:\-1} seconds"\'
.
.fi
.
.P
.
.SS "Proxy with cache in files (for debug API with rate limit)"
get \fBhttp://api\.url/\fR as \fBhttp://localhost:8080/get?url=http://api\.url/\fR
.
.IP "" 4
.
.nf

shell2http \-form \e
    /form \'echo "<html><form action=/get>URL: <input name=url><input type=submit>"\' \e
    /get \'MD5=$(printf "%s" $v_url | md5); cat cache_$MD5 || (curl \-sL $v_url | tee cache_$MD5)\'
.
.fi
.
.IP "" 0
.
.P
.
.SS "Remote sound volume control (Mac OS)"
.
.nf

shell2http /get  \'osascript \-e "output volume of (get volume settings)"\' \e
           /up   \'osascript \-e "set volume output volume (($(osascript \-e "output volume of (get volume settings)")+10))"\' \e
           /down \'osascript \-e "set volume output volume (($(osascript \-e "output volume of (get volume settings)")\-10))"\'
.
.fi
.
.P
.
.SS "Remote control for Vox\.app player (Mac OS)"
.
.nf

shell2http /play_pause \'osascript \-e "tell application \e"Vox\e" to playpause" && echo ok\' \e
           /get_info \'osascript \-e "tell application \e"Vox\e"" \-e "\e"Artist: \e" & artist & \e"\en\e" & \e"Album: \e" & album & \e"\en\e" & \e"Track: \e" & track" \-e "end tell"\'
.
.fi
.
.P
.
.SS "Mock service with JSON API"
.
.nf

curl "http://some\-service/v1/call1" > 1\.json
shell2http \-cgi /call1 \'cat 1\.json\' /call2 \'echo "Content\-Type: application/json\en"; echo "{\e"error\e": \e"ok\e"}"\'
.
.fi
.
.P
.
.SS "Windows example"
Returns value of \fBvar\fR for run in Windows \fBcmd\fR (\fBhttp://localhost:8080/test?var=value123\fR)
.
.IP "" 4
.
.nf

shell2http\.exe \-form /test "echo %v_var%"
.
.fi
.
.IP "" 0
.
.P
.
.SS "With HTTP headers"
Send custom HTTP headers:
.
.IP "" 4
.
.nf

shell2http \-cgi / \'echo "Content\-Type: application/javascript\en"; echo "{\e"error\e": \e"ok\e"}"\'
.
.fi
.
.IP "" 0
.
.P
On Windows:
.
.IP "" 4
.
.nf

shell2http\.exe \-cgi / "echo Content\-Type: application/javascript& echo\.& echo body"
.
.fi
.
.IP "" 0
.
.P
.
.SH "Run from Docker\-container"
Example of \fBtest\.Dockerfile\fR for server for get current date:
.
.IP "" 4
.
.nf

FROM msoap/shell2http
# may be install some alpine packages:
# RUN apk add \-\-no\-cache \.\.\.
CMD ["/date", "date"]
.
.fi
.
.IP "" 0
.
.P
Build and run container:
.
.IP "" 4
.
.nf

docker build \-f test\.Dockerfile \-t date\-server \.
docker run \-\-rm \-p 8080:8080 date\-server
.
.fi
.
.IP "" 0
.
.P
In some cases, such as when multiple shell scripts are running and the main script doesn\'t wait them properly, you can use \fB\-\-init\fR option for \fBdocker run\fR for prevent zombie processes, see Docker run reference \fIhttps://docs\.docker\.com/reference/cli/docker/container/run/#init\fR and about multiple processes in a container \fIhttps://docs\.docker\.com/engine/containers/multi\-service_container/\fR\.
.
.SH "SSL"
Run https server:
.
.IP "" 4
.
.nf

shell2http \-cert=\./cert\.pem \-key=\./key\.pem \.\.\.
.
.fi
.
.IP "" 0
.
.P
Generate self\-signed certificate:
.
.IP "" 4
.
.nf

go run $(go env GOROOT)/src/crypto/tls/generate_cert\.go \-host localhost
.
.fi
.
.IP "" 0
.
.SH "See also"
.
.IP "\(bu" 4
Command\-line tool for creating Telegram bot \- shell2telegram \fIhttps://github\.com/msoap/shell2telegram\fR
.
.IP "\(bu" 4
A http daemon for local development \- devd \fIhttps://github\.com/cortesi/devd\fR
.
.IP "\(bu" 4
Turn any program that uses STDIN/STDOUT into a WebSocket server \- websocketd \fIhttps://github\.com/joewalnes/websocketd\fR
.
.IP "\(bu" 4
The same tool configurable via JSON \- webhook \fIhttps://github\.com/adnanh/webhook\fR
.
.IP "" 0



shell2http.go
package main

import (
	"bytes"
	"context"
	"flag"
	"fmt"
	"html"
	"io"
	"io/ioutil"
	"log"
	"mime/multipart"
	"net"
	"net/http"
	"os"
	"os/exec"
	"regexp"
	"runtime"
	"strconv"
	"strings"
	"time"

	"github.com/mattn/go-shellwords"
	"github.com/msoap/raphanus"
	raphanuscommon "github.com/msoap/raphanus/common"
)

var version = "dev"

const (
	// defaultPort - default port for http-server
	defaultPort = 8080

	// shBasicAuthVar - name of env var for basic auth credentials
	shBasicAuthVar = "SH_BASIC_AUTH"

	// defaultShellPOSIX - shell executable by default in POSIX systems
	defaultShellPOSIX = "sh"

	// defaultShellWindows - shell executable by default in Windows
	defaultShellWindows = "cmd"

	// defaultShellPlan9 - shell executable by default in Plan9
	defaultShellPlan9 = "rc"

	maxHTTPCode            = 1000
	maxMemoryForUploadFile = 65536
)

// indexTmpl - template for index page
const indexTmpl = `<!DOCTYPE html>
<!-- Served by shell2http/%s -->
<html>
<head>
    <title>❯ shell2http</title>
    <style>
    body {
        font-family: sans-serif;
    }
    li {
        list-style-type: none;
    }
    li:before {
        content: "❯";
        padding-right: 5px;
    }
    </style>
</head>
<body>
	<h1>shell2http</h1>
	<ul>
		%s
	</ul>
	Get from: <a href="https://github.com/msoap/shell2http">github.com/msoap/shell2http</a>
</body>
</html>
`

// command - one command
type command struct {
	path       string
	cmd        string
	httpMethod string
	handler    http.HandlerFunc
}

// parsePathAndCommands - get all commands with pathes
func parsePathAndCommands(args []string) ([]command, error) {
	var cmdHandlers []command

	if len(args) < 2 || len(args)%2 == 1 {
		return cmdHandlers, fmt.Errorf("requires a pair of path and shell command")
	}

	pathRe := regexp.MustCompile(`^(?:([A-Z]+):)?(/\S*)$`)
	uniqPaths := map[string]bool{}

	for i := 0; i < len(args); i += 2 {
		path, cmd := args[i], args[i+1]
		if uniqPaths[path] {
			return nil, fmt.Errorf("a duplicate path was detected: %q", path)
		}

		pathParts := pathRe.FindStringSubmatch(path)
		if len(pathParts) != 3 {
			return nil, fmt.Errorf("the path %q must begin with the prefix /, and with optional METHOD: prefix", path)
		}
		cmdHandlers = append(cmdHandlers, command{path: pathParts[2], cmd: cmd, httpMethod: pathParts[1]})

		uniqPaths[path] = true
	}

	return cmdHandlers, nil
}

// getShellAndParams - get default shell and command
func getShellAndParams(cmd string, appConfig Config) (shell string, params []string, err error) {
	shell, params = appConfig.defaultShell, []string{appConfig.defaultShOpt, cmd} // sh -c "cmd"

	// custom shell
	switch {
	case appConfig.shell != appConfig.defaultShell && appConfig.shell != "":
		shell = appConfig.shell
	case appConfig.shell == "":
		cmdLine, err := shellwords.Parse(cmd)
		if err != nil {
			return shell, params, fmt.Errorf("failed to parse %q: %s", cmd, err)
		}

		shell, params = cmdLine[0], cmdLine[1:]
	}

	return shell, params, nil
}

// getShellHandler - get handler function for one shell command
func getShellHandler(appConfig Config, shell string, params []string, cacheTTL raphanus.DB) func(http.ResponseWriter, *http.Request) {
	reStatusCode := regexp.MustCompile(`^\d+`)

	return func(rw http.ResponseWriter, req *http.Request) {
		shellOut, exitCode, err := execShellCommand(appConfig, shell, params, req, cacheTTL)
		if err != nil {
			log.Printf("out: %s, exec error: %s", string(shellOut), err)
		}

		customStatusCode := 0
		outText := string(shellOut)

		if err != nil && !appConfig.showErrors {
			outText = fmt.Sprintf("%s\nexec error: %s", string(shellOut), err)
		} else {
			if appConfig.setCGI {
				var headers map[string]string
				outText, headers = parseCGIHeaders(outText)

				for headerKey, headerValue := range headers {
					switch headerKey {
					case "Status":
						statusParts := reStatusCode.FindAllString(headerValue, -1)
						if len(statusParts) > 0 {
							statusCode, err := strconv.Atoi(statusParts[0])
							if err == nil && statusCode > 0 && statusCode < maxHTTPCode {
								customStatusCode = statusCode
								continue
							}
						}
					case "Location":
						customStatusCode = http.StatusFound
					}

					rw.Header().Set(headerKey, headerValue)
				}
			}
		}

		rw.Header().Set("X-Shell2http-Exit-Code", strconv.Itoa(exitCode))

		if customStatusCode > 0 {
			rw.WriteHeader(customStatusCode)
		} else if exitCode > 0 && appConfig.intServerErr {
			rw.WriteHeader(http.StatusInternalServerError)
		}

		responseWrite(rw, outText)
	}
}

// execShellCommand - execute shell command, returns bytes out and error
func execShellCommand(appConfig Config, shell string, params []string, req *http.Request, cacheTTL raphanus.DB) ([]byte, int, error) {
	if appConfig.cache > 0 {
		if cacheData, err := cacheTTL.GetBytes(req.RequestURI); err != raphanuscommon.ErrKeyNotExists && err != nil {
			log.Printf("get from cache failed: %s", err)
		} else if err == nil {
			// cache hit
			return cacheData, 0, nil // TODO: save exit code in cache
		}
	}

	ctx := req.Context()
	if appConfig.timeout > 0 {
		var cancelFn context.CancelFunc
		ctx, cancelFn = context.WithTimeout(ctx, time.Duration(appConfig.timeout)*time.Second)
		defer cancelFn()
	}
	osExecCommand := exec.CommandContext(ctx, shell, params...) // #nosec

	proxySystemEnv(osExecCommand, appConfig)

	finalizer := func() {}
	if appConfig.setForm {
		var err error
		if finalizer, err = getForm(osExecCommand, req, appConfig.formCheckRe); err != nil {
			log.Printf("parse form failed: %s", err)
		}
	}

	var (
		waitPipeWrite bool
		pipeErrCh     = make(chan error)
		shellOut      []byte
		err           error
	)

	if appConfig.setCGI {
		setCGIEnv(osExecCommand, req, appConfig)

		// get request body data data to stdin of script (if not parse form vars above)
		if (req.Method == "POST" || req.Method == "PUT" || req.Method == "PATCH") && !appConfig.setForm {
			if stdin, pipeErr := osExecCommand.StdinPipe(); pipeErr != nil {
				log.Println("write request body data to shell failed:", pipeErr)
			} else {
				waitPipeWrite = true
				go func() {
					if _, pipeErr := io.Copy(stdin, req.Body); pipeErr != nil {
						pipeErrCh <- pipeErr
						return
					}
					pipeErrCh <- stdin.Close()
				}()
			}
		}
	}

	if appConfig.includeStderr {
		shellOut, err = osExecCommand.CombinedOutput()
	} else {
		var stderrBuf bytes.Buffer
		osExecCommand.Stderr = &stderrBuf
		shellOut, err = osExecCommand.Output()
		if stderrBuf.Len() > 0 {
			log.Printf("stderr: %s", stderrBuf.String())
		}
	}

	if waitPipeWrite {
		if pipeErr := <-pipeErrCh; pipeErr != nil {
			log.Println("write request body data to shell failed:", pipeErr)
		}
	}

	finalizer()

	if appConfig.cache > 0 {
		if cacheErr := cacheTTL.SetBytes(req.RequestURI, shellOut, appConfig.cache); cacheErr != nil {
			log.Printf("set to cache failed: %s", cacheErr)
		}
	}

	exitCode := osExecCommand.ProcessState.ExitCode()

	return shellOut, exitCode, err
}

// setupHandlers - setup http handlers
func setupHandlers(cmdHandlers []command, appConfig Config, cacheTTL raphanus.DB) ([]command, error) {
	resultHandlers := []command{}
	indexLiHTML := []string{}
	existsRootPath := false

	// map[path][http-method]handler
	groupedCmd := map[string]map[string]http.HandlerFunc{}
	cmdsForLog := map[string][]string{}

	for _, row := range cmdHandlers {
		path, cmd := row.path, row.cmd
		shell, params, err := getShellAndParams(cmd, appConfig)
		if err != nil {
			return nil, err
		}

		existsRootPath = existsRootPath || path == "/"

		methodDesc := ""
		if row.httpMethod != "" {
			methodDesc = row.httpMethod + ": "
		}
		indexLiHTML = append(indexLiHTML, fmt.Sprintf(`<li><a href=".%s">%s%s</a> <span style="color: #888">- %s<span></li>`, path, methodDesc, path, html.EscapeString(cmd)))
		cmdsForLog[path] = append(cmdsForLog[path], cmd)

		handler := mwMethodOnly(getShellHandler(appConfig, shell, params, cacheTTL), row.httpMethod)
		if _, ok := groupedCmd[path]; !ok {
			groupedCmd[path] = map[string]http.HandlerFunc{}
		}
		groupedCmd[path][row.httpMethod] = handler
	}

	for path, cmds := range groupedCmd {
		handler, err := mwMultiMethod(cmds)
		if err != nil {
			return nil, err
		}
		resultHandlers = append(resultHandlers, command{
			path:    path,
			handler: handler,
			cmd:     strings.Join(cmdsForLog[path], "; "),
		})
	}

	// --------------
	if appConfig.addExit {
		resultHandlers = append(resultHandlers, command{
			path: "/exit",
			cmd:  "/exit",
			handler: func(rw http.ResponseWriter, _ *http.Request) {
				responseWrite(rw, "Bye...")
				go os.Exit(0)
			},
		})

		indexLiHTML = append(indexLiHTML, fmt.Sprintf(`<li><a href=".%s">%s</a></li>`, "/exit", "/exit"))
	}

	// --------------
	if !appConfig.noIndex && !existsRootPath {
		indexHTML := fmt.Sprintf(indexTmpl, version, strings.Join(indexLiHTML, "\n"))
		resultHandlers = append(resultHandlers, command{
			path: "/",
			cmd:  "index page",
			handler: func(rw http.ResponseWriter, req *http.Request) {
				if req.URL.Path != "/" {
					log.Printf("%s - 404", req.URL.Path)
					http.NotFound(rw, req)
					return
				}

				responseWrite(rw, indexHTML)
			},
		})
	}

	//healthcheck
	resultHandlers = append(resultHandlers, command{
		path: healthcheckPath,
		cmd:  "healthcheck",
		handler: func(rw http.ResponseWriter, _ *http.Request) {
			responseWrite(rw, "ok")
		},
	})

	return resultHandlers, nil
}

// responseWrite - write text to response
func responseWrite(rw io.Writer, text string) {
	if _, err := io.WriteString(rw, text); err != nil {
		log.Printf("print string failed: %s", err)
	}
}

// setCGIEnv - set some CGI variables
func setCGIEnv(cmd *exec.Cmd, req *http.Request, appConfig Config) {
	// set HTTP_* variables
	for headerName, headerValue := range req.Header {
		envName := strings.ToUpper(strings.Replace(headerName, "-", "_", -1))
		if envName == "PROXY" {
			continue
		}
		cmd.Env = append(cmd.Env, fmt.Sprintf("HTTP_%s=%s", envName, headerValue[0]))
	}

	remoteHost, remotePort, err := net.SplitHostPort(req.RemoteAddr)
	if err != nil {
		log.Printf("failed to parse remote address %s: %s", req.RemoteAddr, err)
	}
	CGIVars := [...]struct {
		cgiName, value string
	}{
		{"PATH_INFO", req.URL.Path},
		{"QUERY_STRING", req.URL.RawQuery},
		{"REMOTE_ADDR", remoteHost},
		{"REMOTE_PORT", remotePort},
		{"REQUEST_METHOD", req.Method},
		{"REQUEST_URI", req.RequestURI},
		{"SCRIPT_NAME", req.URL.Path},
		{"SERVER_NAME", appConfig.host},
		{"SERVER_PORT", strconv.Itoa(appConfig.port)},
		{"SERVER_PROTOCOL", req.Proto},
		{"SERVER_SOFTWARE", "shell2http"},
	}

	for _, row := range CGIVars {
		cmd.Env = append(cmd.Env, fmt.Sprintf("%s=%s", row.cgiName, row.value))
	}
}

/*
	parse headers from script output:

Header-name1: value1\n
Header-name2: value2\n
\n
text
*/
func parseCGIHeaders(shellOut string) (string, map[string]string) {
	parts := regexp.MustCompile(`\r?\n\r?\n`).Split(shellOut, 2)
	if len(parts) == 2 {

		headerRe := regexp.MustCompile(`^([^:\s]+):\s*(\S.*)$`)
		headerLines := regexp.MustCompile(`\r?\n`).Split(parts[0], -1)
		headersMap := map[string]string{}

		for _, headerLine := range headerLines {
			headerParts := headerRe.FindStringSubmatch(headerLine)
			if len(headerParts) == 3 {
				headersMap[headerParts[1]] = headerParts[2]
			} else {
				// headers is not valid, return all text
				return shellOut, nil
			}
		}

		return parts[1], headersMap
	}

	// headers don't found, return all text
	return shellOut, nil
}

// getForm - parse form into environment vars, also handle uploaded files
func getForm(cmd *exec.Cmd, req *http.Request, checkFormRe *regexp.Regexp) (func(), error) {
	tempDir := ""
	safeFileNameRe := regexp.MustCompile(`[^\.\w\-]+`)
	finalizer := func() {
		if tempDir != "" {
			if err := os.RemoveAll(tempDir); err != nil {
				log.Println(err)
			}
		}
	}

	if err := req.ParseForm(); err != nil {
		return finalizer, err
	}

	if isMultipartFormData(req.Header) {
		if err := req.ParseMultipartForm(maxMemoryForUploadFile); err != nil {
			return finalizer, err
		}
	}

	for key, values := range req.Form {
		if checkFormRe != nil {
			checkedValues := []string{}
			for _, v := range values {
				if checkFormRe.MatchString(v) {
					checkedValues = append(checkedValues, v)
				}
			}
			values = checkedValues
		}
		if len(values) == 0 {
			continue
		}

		value := strings.Join(values, ",")
		cmd.Env = append(cmd.Env, fmt.Sprintf("%s=%s", "v_"+key, value))
	}

	// handle uploaded files, save all to temporary files and set variables filename_XXX, filepath_XXX
	if req.MultipartForm != nil {
		for key, value := range req.MultipartForm.File {
			if len(value) == 1 {
				var (
					uplFile     multipart.File
					outFile     *os.File
					err         error
					reqFileName = value[0].Filename
				)

				errCreate := errChain(func() error {
					uplFile, err = value[0].Open()
					return err
				}, func() error {
					tempDir, err = ioutil.TempDir("", "shell2http_")
					return err
				}, func() error {
					prefix := safeFileNameRe.ReplaceAllString(reqFileName, "")
					outFile, err = ioutil.TempFile(tempDir, prefix+"_")
					return err
				}, func() error {
					_, err = io.Copy(outFile, uplFile)
					return err
				})

				errClose := errChainAll(func() error {
					if uplFile != nil {
						return uplFile.Close()
					}
					return nil
				}, func() error {
					if outFile != nil {
						return outFile.Close()
					}
					return nil
				})
				if errClose != nil {
					return finalizer, errClose
				}

				if errCreate != nil {
					return finalizer, errCreate
				}

				cmd.Env = append(cmd.Env, fmt.Sprintf("%s=%s", "filepath_"+key, outFile.Name()))
				cmd.Env = append(cmd.Env, fmt.Sprintf("%s=%s", "filename_"+key, reqFileName))
			}
		}
	}

	return finalizer, nil
}

// isMultipartFormData - check header for multipart/form-data
func isMultipartFormData(headers http.Header) bool {
	if contentType, ok := headers["Content-Type"]; ok && len(contentType) == 1 && strings.HasPrefix(strings.ToLower(contentType[0]), "multipart/form-data;") {
		return true
	}

	return false
}

// proxySystemEnv - proxy some system vars
func proxySystemEnv(cmd *exec.Cmd, appConfig Config) {
	varsNames := []string{"PATH", "HOME", "LANG", "USER", "TMPDIR"}

	if runtime.GOOS == "windows" {
		varsNames = append(varsNames, "USERNAME", "USERPROFILE", "HOMEDRIVE", "HOMEPATH", "TEMP", "TMP", "PATHEXT", "COMSPEC", "OS")
	}

	if appConfig.exportVars != "" {
		varsNames = append(varsNames, strings.Split(appConfig.exportVars, ",")...)
	}

	for _, envRaw := range os.Environ() {
		env := strings.SplitN(envRaw, "=", 2)
		if env[0] != shBasicAuthVar {
			if appConfig.exportAllVars {
				cmd.Env = append(cmd.Env, envRaw)
			} else {
				for _, envVarName := range varsNames {
					if strings.ToUpper(env[0]) == envVarName {
						cmd.Env = append(cmd.Env, envRaw)
					}
				}
			}
		}
	}
}

// errChain - handle errors on few functions
func errChain(chainFuncs ...func() error) error {
	for _, fn := range chainFuncs {
		if err := fn(); err != nil {
			return err
		}
	}

	return nil
}

// errChainAll - handle errors on few functions, exec all func and returns the first error
func errChainAll(chainFuncs ...func() error) error {
	var resErr error
	for _, fn := range chainFuncs {
		if err := fn(); err != nil {
			resErr = err
		}
	}

	return resErr
}

func main() {
	appConfig, err := getConfig()
	if err != nil {
		log.Fatal(err)
	}

	cmdHandlers, err := parsePathAndCommands(flag.Args())
	if err != nil {
		log.Fatalf("failed to parse arguments: %s", err)
	}

	var cacheTTL raphanus.DB
	if appConfig.cache > 0 {
		cacheTTL = raphanus.New()
	}

	cmdHandlers, err = setupHandlers(cmdHandlers, *appConfig, cacheTTL)
	if err != nil {
		log.Fatal(err)
	}
	for _, handler := range cmdHandlers {
		handlerFunc := handler.handler
		if len(appConfig.auth.users) > 0 {
			handlerFunc = mwBasicAuth(handlerFunc, appConfig.auth)
		}
		if appConfig.oneThread {
			handlerFunc = mwOneThread(handlerFunc)
		}
		handlerFunc = mwLogging(mwCommonHeaders(handlerFunc))

		http.HandleFunc(handler.path, handlerFunc)
		log.Printf("register: %s (%s)\n", handler.path, handler.cmd)
	}

	listener, err := net.Listen("tcp", net.JoinHostPort(appConfig.host, strconv.Itoa(appConfig.port)))
	if err != nil {
		log.Fatal(err)
	}

	log.Printf("listen %s\n", appConfig.readableURL(listener.Addr()))

	if len(appConfig.cert) > 0 && len(appConfig.key) > 0 {
		log.Fatal(http.ServeTLS(listener, nil, appConfig.cert, appConfig.key))
	} else {
		log.Fatal(http.Serve(listener, nil))
	}
}
